hphp/runtime/vm/jit/relocation-arm.cpp

```diff
@@ -318,6 +320,55 @@ bool relocatePCRelative(Env& env, Instruction* src, Instruction* dest,
         destCount += (env.destBlock.frontier() - destAddr) / kInstructionSize;
         isRelative = false;
       }
+    } else if (src->IsCompareBranch()) {
+      imm >>= vixl::kInstructionSizeLog2;
+      if (!is_int19(imm) || env.far.count(src)) {
```

Comment:
Consider writing a helper `imm_fits_instr(int imm, const Instruction* instr)` which does the work of switching on the instruction kind, shifting the immediate, and dispatching to the appropriate size check.  There are a few places where you're hand-rolling this dispatch that could be papered over, such as...